### PR TITLE
SAAS-3083: Added Kendra Index Encryption Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/kendra/kendraIndexEncrypted.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            default: 'awscmk'
+        }
+    },
+
+    run: function(cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var regions = helpers.regions(settings);
+
+
```
Comment (sadeed12345) [inline plugins/aws/kendra/kendraIndexEncrypted.js:28]:
```suggestion
```suggestion
```
```

File: plugins/aws/kendra/kendraIndexEncrypted.spec.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+];
+
+const describeIndex = [
+    {
+    "Name": "sadeed1",
+    "Id": "9280dadd-5d45-4f9c-a105-896e5b230c05",
+    "Edition": "DEVELOPER_EDITION",
+    "RoleArn": "arn:aws:iam::101363889637:role/service-role/AmazonKendra-us-east-1-role1",
+    "ServerSideEncryptionConfiguration": {
+        "KmsKeyId": "arn:aws:kms:us-east-1:101363889637:key/ad013a33-b01d-4d88-ac97-127399c18b3e"
```
Comment (sadeed12345) [inline plugins/aws/kendra/kendraIndexEncrypted.spec.js:30]:
```suggestion
        "KmsKeyId": "arn:aws:kms:us-east-1:000111222333:key/ad013a33-b01d-4d88-ac97-127399c18b3e"
```

File: helpers/aws/regions_gov.js [hunk 3]
```diff
@@ -80,6 +80,8 @@ module.exports = {
     codebuild: regions,
     mq: regions,
     glacier: regions,
+    kendra:  ['us-east-1', 'us-east-2', 'us-west-2','ap-southeast-1',
```
Comment (dipsubha06) [inline helpers/aws/regions_gov.js:83]: @AkhtarAmir is this okay or it should be just regions, can you please verify this.

File: helpers/aws/regions.js [hunk 4]
```diff
@@ -112,6 +112,9 @@ module.exports = {
     codebuild: regions,
     mq: regions,
     glacier: regions,
+    kendra:  ['us-east-1', 'us-east-2', 'us-west-2','ap-southeast-1',
+        'ap-southeast-2', 'ca-central-1', 'eu-west-1'],
+
```
Comment (AkhtarAmir) [inline helpers/aws/regions.js:117]:
```suggestion
```

File: helpers/aws/regions.js [hunk 5]
```diff
@@ -112,6 +112,8 @@ module.exports = {
     codebuild: regions,
     mq: regions,
     glacier: regions,
+    kendra:  ['us-east-1', 'us-east-2', 'us-west-2','ap-southeast-1',
+        'ap-southeast-2', 'ca-central-1', 'eu-west-1'],
```
Comment (AkhtarAmir) [inline helpers/aws/regions.js:116]:
```suggestion
    kendra:  ['us-east-1', 'us-east-2', 'us-west-2','ap-southeast-1', 'ap-southeast-2', 'ca-central-1', 'eu-west-1'],
```

File: plugins/aws/kendra/kendraIndexEncrypted.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Kendra Index Encrypted',
+    category: 'Kendra',
+    domain: 'Databases',
+    description: 'Ensure that the Kendra Index is Encrypted',
```
Comment (AkhtarAmir) [inline plugins/aws/kendra/kendraIndexEncrypted.js:8]:
```suggestion
    description: 'Ensure that the Kendra index is encrypted using desired encryption level.',
```

File: plugins/aws/kendra/kendraIndexEncrypted.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Kendra Index Encrypted',
+    category: 'Kendra',
+    domain: 'Databases',
+    description: 'Ensure that the Kendra index is encrypted using desired encryption level.',
+    more_info: 'Amazon Kendra encrypts your data with AWS-manager keys by default.' +
+               'Encrypt your files using customer-managed keys in order to gain more granular control over encryption/decryption process.',
```
Comment (AkhtarAmir) [inline plugins/aws/kendra/kendraIndexEncrypted.js:10]:
```suggestion
    more_info: 'Amazon Kendra encrypts your data at rest with AWS-manager keys by default.' +
               'Use customer-managed keys instead in order to gain more granular control over encryption/decryption process.',
```

File: plugins/aws/kendra/kendraIndexEncrypted.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+                var resource = `arn:${awsOrGov}:kendra:${accountId}:index/${index.Name}`;
+
+                var describeIndex = helpers.addSource(cache, source,
+                    ['kendra', 'describeIndex', region, index.Id]);
+                    
+                if (!describeIndex || describeIndex.err || !describeIndex.data ) {
+                    helpers.addResult(results, 3,
+                        `Unable to get Kendra Indices description: ${helpers.addError(describeIndex)}`,
+                        region, resource);
+                }
```
Comment (AkhtarAmir) [inline plugins/aws/kendra/kendraIndexEncrypted.js:74]:
```suggestion
                if (!describeIndex || describeIndex.err || !describeIndex.data ) {
                    helpers.addResult(results, 3,
                        `Unable to get Kendra Indices description: ${helpers.addError(describeIndex)}`,
                        region, resource);
                    continue;
                } 
```

File: plugins/aws/kendra/kendraIndexEncrypted.js [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                            region, kmsKeyId);
+                        continue;
+                    }
+
+                    currentEncryptionLevel = helpers.getEncryptionLevel(describeKey.data.KeyMetadata, helpers.ENCRYPTION_LEVELS);
+                } else {
+
+                    currentEncryptionLevel=2; //awskms
+                }
+
```
Comment (AkhtarAmir) [inline plugins/aws/kendra/kendraIndexEncrypted.js:98]:
```suggestion
                } else {
                    currentEncryptionLevel = 2; //awskms
                }

```

File: plugins/aws/kendra/kendraIndexEncrypted.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+                var currentEncryptionLevelString = helpers.ENCRYPTION_LEVELS[currentEncryptionLevel];
+
+                if (currentEncryptionLevel >= desiredEncryptionLevel) {
+                    helpers.addResult(results, 0,
+                        `Kendra Indices is encrypted with ${currentEncryptionLevelString} \
```
Comment (AkhtarAmir) [inline plugins/aws/kendra/kendraIndexEncrypted.js:102]:
```suggestion
                        `Kendra index is encrypted with ${currentEncryptionLevelString} \
```

File: plugins/aws/kendra/kendraIndexEncrypted.js [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+                        which is greater than or equal to the desired encryption level ${config.desiredEncryptionLevelString}`,
+                        region, resource);
+                } else {
+                    helpers.addResult(results, 2,
+                        `Kendra Indices is encrypted with ${currentEncryptionLevelString} \
```
Comment (AkhtarAmir) [inline plugins/aws/kendra/kendraIndexEncrypted.js:107]:
```suggestion
                        `Kendra index is encrypted with ${currentEncryptionLevelString} \
```

File: plugins/aws/kendra/kendraIndexEncrypted.js [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+                        region, resource);
+                    continue;
+                } 
+
+                if (describeIndex.data.ServerSideEncryptionConfiguration &&
+                    describeIndex.data.ServerSideEncryptionConfiguration.KmsKeyId) {
+
+                    var kmsKeyId = describeIndex.data.ServerSideEncryptionConfiguration.KmsKeyId;
+                    var keyId = kmsKeyId.split('/')[1] ? kmsKeyId.split('/')[1] : kmsKeyId;
+
```
Comment (AkhtarAmir) [inline plugins/aws/kendra/kendraIndexEncrypted.js:82]:
```suggestion
                    describeIndex.data.ServerSideEncryptionConfiguration.KmsKeyId) {
                    var kmsKeyId = describeIndex.data.ServerSideEncryptionConfiguration.KmsKeyId;
                    var keyId = kmsKeyId.split('/')[1] ? kmsKeyId.split('/')[1] : kmsKeyId;

```

File: plugins/aws/kendra/kendraIndexEncrypted.js [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+            for (let index of listIndices.data) {
+                var resource = `arn:${awsOrGov}:kendra:${accountId}:index/${index.Name}`;
+
+                var describeIndex = helpers.addSource(cache, source,
+                    ['kendra', 'describeIndex', region, index.Id]);
+                    
+                if (!describeIndex || describeIndex.err || !describeIndex.data ) {
+                    helpers.addResult(results, 3,
+                        `Unable to get Kendra Indices description: ${helpers.addError(describeIndex)}`,
```
Comment (AkhtarAmir) [inline plugins/aws/kendra/kendraIndexEncrypted.js:72]:
```suggestion
                        `Unable to get Kendra index description: ${helpers.addError(describeIndex)}`,
```

File: plugins/aws/kendra/kendraIndexEncrypted.js [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+                ['kms', 'listKeys', region]);
+
+            if (!listKeys || listKeys.err || !listKeys.data) {
+                helpers.addResult(results, 3,
+                    `Unable to list KMS keys: ${helpers.addError(listKeys)}`, region);
+                return rcb();
+            }
+
+            for (let index of listIndices.data) {
+                var resource = `arn:${awsOrGov}:kendra:${accountId}:index/${index.Name}`;
```
Comment (AkhtarAmir) [inline plugins/aws/kendra/kendraIndexEncrypted.js:65]:
```suggestion
                var resource = `arn:${awsOrGov}:kendra:${region}:${accountId}:index/${index.Name}`;
```

**Pull request merged**